src/Group.ts

```diff
@@ -0,0 +1,321 @@
+     * group names exactly matching groups().
+     */
+    get tasks(): Task[] {
+        return this._tasks;
+    }
+
+    /**
+     * A markdown-format representation of all the tasks in this group.
+     *
+     * Useful for testing.
+     */
+    public tasksAsStringOfLines(): string {
+        let output = '';
+        for (const task of this.tasks) {
+            output += task.toFileLineString() + '\n';
+        }
+        return output;
+    }
+
+    /**
+     * A human-readable representation of this task group, including names
+     * and headings that should be displayed.
+     *
+     * Note that this is used in snapshot testing, so if the format is
+     * changed, the snapshots will need to be updated.
+     */
+    public toString(): string {
+        let output = '\n';
+        output += `Group names: [${this.groups}]\n`;
+
+        for (const heading of this.groupHeadings) {
+            // These headings mimic the behaviour of QueryRenderer,
+            // which uses 'h4', 'h5' and 'h6' for nested groups.
+            const headingPrefix = '#'.repeat(4 + heading.nestingLevel);
+            output += `${headingPrefix} ${heading.name}\n`;
+        }
+
+        output += this.tasksAsStringOfLines();
+        return output;
+    }
+
+    private readonly _groups: string[];
+    private readonly _groupHeadings: GroupHeading[];
+    private readonly _tasks: Task[];
+}
+
+/**
+ * TaskGroup stores all the groups of tasks generated by any 'group by'
+ * instructions in the task block.
+ */
+export class TaskGroups {
+    /**
+     * Constructor for TaskGroups
+     * @param {Grouping[]} groups - 0 or more Grouping values,
+     *                              1 per 'group by' line in the task query block
+     * @param {Task[]} tasks - 0 more more Task objects, with all the tasks
+     *                         matching the query, already in sort order
+     */
+    constructor(groups: Grouping[], tasks: Task[]) {
+        const initialGroups = new IntermediateTaskGroups(groups, tasks);
+        this.addTasks(initialGroups);
+    }
+
+    private addTasks(initialGroups: IntermediateTaskGroups) {
+        // Get the headings
+        const grouper = new GroupHeadings(initialGroups.groups);
+
+        // Build a container of all the groups
+        for (const [groups, tasks] of initialGroups.groups) {
+            const groupHeadings = grouper.getHeadingsForTaskGroup(groups);
+            const taskGroup = new TaskGroup(groups, groupHeadings, tasks);
+            this.add(taskGroup);
+        }
+    }
+
+    private add(taskGroup: TaskGroup) {
+        this._groups.push(taskGroup);
+    }
+
+    /**
+     * All the tasks matching the query, grouped together, and in the order
+     * that they should be displayed.
+     */
+    get groups(): TaskGroup[] {
+        return this._groups;
+    }
+
+    /**
+     * The total number of tasks matching the query - after applying any 'limit' instruction.
```

Comment:
Indeed. This class doesn't apply any limiting. I was trying to convey that that in this comment.
But since it's obviously confusing, I've removes the reference to the the limit - in b1d47a6d928abda2af9e2e2c8df390027fb28e50